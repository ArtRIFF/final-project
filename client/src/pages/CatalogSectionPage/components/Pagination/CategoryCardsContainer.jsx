import React from "react";
import LoadingSpinner from "../LoadingSpinner/LoadingSpinner";
import CategorySectionCard from "../../../../components/CatalogSection/CategorySectionCard";
import "./CategoryCardsContainer.scss";

const CategoryCardsContainer = ({
  loading,
  allCollectionArray,
  filterSearchingResults,
}) => {
  if (!allCollectionArray) {
    return <LoadingSpinner />;
  }


    // if (allCollectionArray) {
    //   return (
    //     <div
    //       style={{
    //         fontFamily: "Libre Bodoni Bold",
    //         fontWeight: "200",
    //         fontSize: "32px",
    //         textAlign: "center",
    //         paddingTop: "180px",
    //       }}
    //     >
    //       Sorry, no items matching your search criteria
    //     </div>
    //   );
    // }

  return (
      <div className="categoryCards__grid">
        {allCollectionArray&&allCollectionArray.map((card) => (
          <div className="CategorySectionCard__wrapper" key={Math.random()}>
            <CategorySectionCard product={card} />
          </div>
        ))}
      </div>
  );
};

export default CategoryCardsContainer;
