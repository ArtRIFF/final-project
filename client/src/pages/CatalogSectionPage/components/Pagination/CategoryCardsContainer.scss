@import "../../../../global_scss/vars.scss";
@import "../../../../global_scss/mixins.scss";

.CategorySectionCard__wrapper {
  height: 100%;
  text-align: center;
  .category-card {
    margin-right: 0;
    height: 100%;
  }
}

.categoryCards__grid {
  justify-content: center;
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  grid-template-rows: 1fr;
  align-items: start;
  gap: 10px;

  @media (max-width: $notebook-size) {
    grid-template-columns: repeat(2, 1fr);
    justify-items: center;
    .category-card {
      width: 100%;
      height: 100%;
      max-width: 230px;
      margin: 0;
    }
  }
  @media (max-width: $mobile-size) {
    grid-template-columns: repeat(1, 1fr);
  }
}
