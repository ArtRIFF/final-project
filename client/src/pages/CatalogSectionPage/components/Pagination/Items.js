import React from "react";

const Items = ({ items, loading }) => {
  if (loading) {
    return <h2>loading...</h2>;
  }

  return (
    <ul>
      {items.map((item, index) => (
        <li>{item.name}</li>
        //зображення товару та решта пропсів
      ))}
    </ul>
  );
};

export default Items;
