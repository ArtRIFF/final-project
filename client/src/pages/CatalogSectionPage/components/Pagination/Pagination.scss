.pagination-container {
  display: flex;
  justify-content: center;
  align-items: center;
  width: 100%;
  font-weight: 500;
  font-style: 15px;
}

.pagination-container a {
  display: flex;
  justify-content: center;
  align-items: center;
  color: black;
  height: 40px;
  width: 38px;
  text-decoration: none;
  transition: background-color .2s;
  border: 1px solid #ddd;
  cursor: pointer;
}

.pagination-container a.active {
  background-color: #E36709;
  color: white;
  border: 1px solid #e55209;
}

.pagination-container a.disabled {
  opacity: 0.2;
}

.pagination-container a:hover:not(.active) {background-color: rgb(238, 238, 238);}