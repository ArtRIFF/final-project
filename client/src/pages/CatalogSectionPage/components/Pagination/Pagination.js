import React, { useState, useEffect } from "react";
import "./Pagination.scss";

const Pagination = ({ itemsPerPage, totalItems, setCurrentPage }) => {
  const pageNumbers = [];

  const [currentButton, setCurrentButton] = useState(1);
  const [arrayOfCurrentButtons, setArrayOfCurrentButtons] = useState([]);

  for (let i = 1; i <= Math.ceil(totalItems / itemsPerPage); i++) {
    pageNumbers.push(i);
  }

  useEffect(() => {
    let tempNumberOfPages = [...arrayOfCurrentButtons];

    let dotsInitial = "...";
    let dotsLeft = "... ";
    let dotsRight = " ...";

    if (pageNumbers.length < 6) {
      tempNumberOfPages = pageNumbers;
    } else if (currentButton >= 1 && currentButton <= 3) {
      tempNumberOfPages = [1, 2, 3, 4, dotsInitial, pageNumbers.length];
    } else if (currentButton === 4) {
      const sliced = pageNumbers.slice(0, 5);
      tempNumberOfPages = [...sliced, dotsInitial, pageNumbers.length];
    } else if (currentButton > 4 && currentButton < pageNumbers.length - 2) {
      const sliced1 = pageNumbers.slice(currentButton - 2, currentButton);
      const sliced2 = pageNumbers.slice(currentButton, currentButton + 1);
      tempNumberOfPages = [
        1,
        dotsLeft,
        ...sliced1,
        ...sliced2,
        dotsRight,
        pageNumbers.length,
      ];
    } else if (currentButton > pageNumbers.length - 3) {
      const sliced = pageNumbers.slice(pageNumbers.length - 4);
      tempNumberOfPages = [1, dotsLeft, ...sliced];
    } else if (currentButton === dotsInitial) {
      setCurrentButton(
        arrayOfCurrentButtons[arrayOfCurrentButtons.length - 3] + 1
      );
    } else if (currentButton === dotsRight) {
      setCurrentButton(arrayOfCurrentButtons[3] + 2);
    } else if (currentButton === dotsLeft) {
      setCurrentButton(arrayOfCurrentButtons[3] - 2);
    }

    setArrayOfCurrentButtons(tempNumberOfPages);
    setCurrentPage(currentButton);
  }, [currentButton]);

  return (
    <div>
      <ul className="pagination-container">
        <a
          href="javascript:void(0)"
          // href="#"
          className={`${currentButton === 1 ? "disabled" : ""}`}
          onClick={() =>
            setCurrentButton((prev) => (prev <= 1 ? prev : prev - 1))
          }
        >
          Prev
        </a>
        {arrayOfCurrentButtons.map((page) => (
          <li key={page}>
            <a
              key={page}
              href="javascript:void(0);"
              // href="#"
              className={`${currentButton === page ? "active" : ""}`}
              onClick={() => {
                setCurrentButton(page);
              }}
            >
              {" "}
              {page}{" "}
            </a>
          </li>
        ))}
        <a
          href="javascript:void(0)"
          // href="#"
          className={`${
            currentButton === pageNumbers.length ? "disabled" : ""
          }`}
          onClick={() =>
            setCurrentButton((prev) =>
              prev >= pageNumbers.length ? prev : prev + 1
            )
          }
        >
          Next
        </a>
      </ul>
    </div>
  );
};

export default Pagination;
