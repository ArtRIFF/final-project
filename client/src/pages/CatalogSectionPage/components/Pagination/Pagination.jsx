import React, { useState, useEffect } from "react";
import { Link } from "react-router-dom";
import "./Pagination.scss";

const Pagination = ({
  itemsPerPage,
  totalItems,
  setCurrentPage,
  allCollectionArrayIsFiltered,
  paginationRequest
}) => {
  const pageNumbers = [];


  const [currentButton, setCurrentButton] = useState(1);
  const [arrayOfCurrentButtons, setArrayOfCurrentButtons] = useState([]);

  for (let i = 1; i <= Math.ceil(totalItems / itemsPerPage); i++) {
    pageNumbers.push(i);
  }

  useEffect(() => {
    if (currentButton !== 1) {
      paginationRequest(currentButton);   
    }
  }, [currentButton]);


  useEffect(() => {
    let tempNumberOfPages = [...arrayOfCurrentButtons];

    let dotsInitial = "...";
    let dotsLeft = "... ";
    let dotsRight = " ...";

    if (pageNumbers.length < 6) {
      tempNumberOfPages = pageNumbers;
    } else if (currentButton >= 1 && currentButton <= 3) {
      tempNumberOfPages = [1, 2, 3, 4, dotsInitial, pageNumbers.length];
    } else if (currentButton === 4) {
      const sliced = pageNumbers.slice(0, 5);
      tempNumberOfPages = [...sliced, dotsInitial, pageNumbers.length];
    } else if (currentButton > 4 && currentButton < pageNumbers.length - 2) {
      const sliced1 = pageNumbers.slice(currentButton - 2, currentButton);
      const sliced2 = pageNumbers.slice(currentButton, currentButton + 1);
      tempNumberOfPages = [
        1,
        dotsLeft,
        ...sliced1,
        ...sliced2,
        dotsRight,
        pageNumbers.length,
      ];
    } else if (currentButton > pageNumbers.length - 3) {
      const sliced = pageNumbers.slice(pageNumbers.length - 4);
      tempNumberOfPages = [1, dotsLeft, ...sliced];
    } else if (currentButton === dotsInitial) {
      setCurrentButton(
        arrayOfCurrentButtons[arrayOfCurrentButtons.length - 3] + 1
      );
    } else if (currentButton === dotsRight) {
      setCurrentButton(arrayOfCurrentButtons[3] + 2);
    } else if (currentButton === dotsLeft) {
      setCurrentButton(arrayOfCurrentButtons[3] - 2);
    }
    setArrayOfCurrentButtons(tempNumberOfPages);
    setCurrentPage(currentButton);
  }, [currentButton, totalItems]);


  useEffect(() => {
    if (allCollectionArrayIsFiltered === true) {
      if (currentButton > arrayOfCurrentButtons.length / itemsPerPage) {
        setCurrentButton(1);
      }
    }
  }, [allCollectionArrayIsFiltered]);

  return (
    <div style={{marginTop:"30px"}}>
      <ul className="pagination-container">
        <Link
          href="javascript:void(0)"
          className={`${currentButton === 1 ? "disabled" : ""}`}
          onClick={() =>
            setCurrentButton((prev) => (prev <= 1 ? prev : prev - 1))
          }
        >
          Prev
        </Link>
        {arrayOfCurrentButtons.map((page) => (
          <li key={page}>
            <Link
              key={page}
              href="javascript:void(0);"
              // href="#"
              className={`${currentButton === page ? "active" : ""}`}
              onClick={() => {
                setCurrentButton(page);
              }}
            >
              {" "}
              {page}{" "}
            </Link>
          </li>
        ))}
        <Link
          href="javascript:void(0)"
          // href="#"
          className={`${
            currentButton === pageNumbers.length ? "disabled" : ""
          }`}
          onClick={() =>
            setCurrentButton((prev) =>
              prev >= pageNumbers.length ? prev : prev + 1
            )
          }
        >
          Next
        </Link>
      </ul>
    </div>
  );
};

export default Pagination;
