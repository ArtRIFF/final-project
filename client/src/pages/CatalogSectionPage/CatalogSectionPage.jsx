import "./style.scss";
import Breadcrumbs from "./components/Breadcrumbs/Breadcrumbs";
import CategoryFilter from "./components/CategoryFilter/CategoryFilter";
import AsideFilter from "./components/AsideFilter/AsideFilter";
import Items from "./components/Pagination/Items";
import Pagination from "./components/Pagination/Pagination";

import { useDispatch, useSelector } from "react-redux";
import { useState, useEffect } from "react";

import { selectorAllCollectionProduct } from "../../store/selectors";
import { fetchAllCollectionProduct } from "../../store/actions";

const CatalogSectionPage = ({ alreadyFilteredArray }) => {
  const dispatch = useDispatch();

  const [items, setitems] = useState([
    11, 124, 1244, 1241, 12, 12, 1, 12, 12, 12, 3, 4, 5, 6, 7, 3, 12, 12, 12,
    124, 15, 12, 53735, 12312,
  ]);
  const [loading, setLoading] = useState(false);
  const [currentPage, setCurrentPage] = useState(1);
  const [itemsPerPage] = useState(5);
  const [filtredArray, setFiltredArray] = useState(alreadyFilteredArray);
  const lastItemIndex = currentPage * itemsPerPage;
  const firstItemIndex = lastItemIndex - itemsPerPage;
  const currentItem = items.slice(firstItemIndex, lastItemIndex);

  useEffect(() => {
    const getItems = async () => {
      setLoading(true);
      // const res = await axios.get("https://final-backend-new.onrender.com"); //адреса сторінки на бекенді для отримання елементів сторінки
      // setitems(res.data);
      setLoading(false);
    };
    getItems();
  }, []);
  const [showAsideFilter, setModalRender] = useState(false);
  const allCollectionArray = useSelector(selectorAllCollectionProduct);

  const callAsideFilter = () => {
    setModalRender(true);
  };

  useEffect(() => {
    dispatch(fetchAllCollectionProduct());
  }, []);

  const filterRequest = (array) => {
    setFiltredArray(array);
  };

  const hideAsideFilter = (event) => {
    const isFilterElement = !!event.target.closest(
      ".asideFilter-wrapper--show"
    );
    const isCallButton = !!event.target.closest(".category-filter--btn");

    if (event && !isFilterElement && !isCallButton) {
      setModalRender(false);
    }
  };

  const array = Array.isArray(filtredArray)
    ? filtredArray.length
    : allCollectionArray.length;

  return (
    <div className="container" onClick={hideAsideFilter}>
      <div className="grid-wrapper">
        <div className="breadcrumbs-wrapper">
          <Breadcrumbs />
        </div>
        <div className="catalogPageImg-wrapper">
          <img
            src="img/catalogSectionPage/CategorySectionMainImg.jpg"
            alt="Category Section Main Imgage"
          />
        </div>
        <aside
          className={`${showAsideFilter
            ? "asideFilter-wrapper--show"
            : "asideFilter-wrapper"
            }`}
        >
          <AsideFilter
            allCollectionArray={allCollectionArray}
            filterRequest={filterRequest}
          />
        </aside>
        <div className="filter-wrapper">
          <CategoryFilter onClickFunc={callAsideFilter} setResult={array} />
        </div>
        <div
          style={{
            backgroundColor: "rgba(100, 85, 45, 0.5)",
            width: "850px",
            height: "510px",
          }}
          className="categoryCards-wrapper"
        >
          {!Array.isArray(filtredArray)
            ? allCollectionArray.map((item, i) => {
              const { name, insertNumber, alt, bestseller, newProduct } = item;
              return (
                <p key={i}>
                  {i} Product:{name} insertNumber:{insertNumber} alt:{alt} bestseller:
                  {bestseller} newProduct:{newProduct}
                </p>
              );
            })
            : filtredArray.map((item, i) => {
              const { name, statusProduct, insertNumber, collectionName, categories, metal } = item;
              return (
                <p key={i}>
                  {i} name:{name} statusProduct:{statusProduct} categories:{categories}
                </p>
              );
            })}
        </div>
        <div
          style={{ backgroundColor: "grey", width: "396px", height: "88px" }}
          className="paginnation-wrapper"
        >
          <Items items={currentItem} loading={loading} />
          <Pagination itemsPerPage={itemsPerPage} totalItems={items.length} />
        </div>
      </div>
    </div>
  );
};

CatalogSectionPage.defaultProps = {
  alreadyFilteredArray: [],
};

export default CatalogSectionPage;
