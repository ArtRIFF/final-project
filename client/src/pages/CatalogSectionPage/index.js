import CatalogSectionPage from "./CatalogSectionPage";
export default CatalogSectionPage;
