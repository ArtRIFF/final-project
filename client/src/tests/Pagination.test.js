import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import { setCurrentPage } from "../pages/CatalogSectionPage/CatalogSectionPage";
import Pagination from "../pages/CatalogSectionPage/components/Pagination/Pagination";

describe("Pagination renders correctly", () => {
  it("Pagination with 0 elements", () => {
    const view = render(
      <BrowserRouter>
        <Pagination
          itemsPerPage={12}
          totalItems={0}
          setCurrentPage={setCurrentPage}
          allCollectionArrayIsFiltered={false}
        ></Pagination>
        ;
      </BrowserRouter>
    );
    expect(view).toMatchSnapshot();
  });

  it("Pagination with 24 elements", () => {
    const view = render(
      <BrowserRouter>
        <Pagination
          itemsPerPage={12}
          totalItems={24}
          setCurrentPage={setCurrentPage}
          allCollectionArrayIsFiltered={false}
        ></Pagination>
        ;
      </BrowserRouter>
    );
    expect(view).toMatchSnapshot();
  });
});
