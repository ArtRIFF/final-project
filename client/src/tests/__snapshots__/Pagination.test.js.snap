// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly Pagination with 0 elements 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="margin-top: 30px;"
      >
        <ul
          class="pagination-container"
        >
          <a
            class="disabled"
            href="/"
          >
            Prev
          </a>
          <a
            class=""
            href="/"
          >
            Next
          </a>
        </ul>
      </div>
      ;
    </div>
  </body>,
  "container": <div>
    <div
      style="margin-top: 30px;"
    >
      <ul
        class="pagination-container"
      >
        <a
          class="disabled"
          href="/"
        >
          Prev
        </a>
        <a
          class=""
          href="/"
        >
          Next
        </a>
      </ul>
    </div>
    ;
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`Pagination renders correctly Pagination with 24 elements 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="margin-top: 30px;"
      >
        <ul
          class="pagination-container"
        >
          <a
            class="disabled"
            href="/"
          >
            Prev
          </a>
          <li>
            <a
              class="active"
              href="/"
            >
               
              1
               
            </a>
          </li>
          <li>
            <a
              class=""
              href="/"
            >
               
              2
               
            </a>
          </li>
          <a
            class=""
            href="/"
          >
            Next
          </a>
        </ul>
      </div>
      ;
    </div>
  </body>,
  "container": <div>
    <div
      style="margin-top: 30px;"
    >
      <ul
        class="pagination-container"
      >
        <a
          class="disabled"
          href="/"
        >
          Prev
        </a>
        <li>
          <a
            class="active"
            href="/"
          >
             
            1
             
          </a>
        </li>
        <li>
          <a
            class=""
            href="/"
          >
             
            2
             
          </a>
        </li>
        <a
          class=""
          href="/"
        >
          Next
        </a>
      </ul>
    </div>
    ;
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
